File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/core/Source.kt
package ca.logaritm.dezel.core

import android.content.Context
import ca.logaritm.dezel.networking.RemoteFileLoader
import java.io.IOException

/**
 * A context source file.
 * @class Source
 * @since 0.7.0
 */
open class Source(context: Context, location: String, category: Category) {

	//--------------------------------------------------------------------------
	// Enum
	//--------------------------------------------------------------------------

	/**
	 * Source categories.
	 * @enum Category
	 * @since 0.7.0
	 */
	public enum class Category {
		STYLE,
		SCRIPT
	}

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * The source's context.
	 * @property context
	 * @since 0.7.0
	 */
	public var context: Context
		private set

	/**
	 * The source's category.
	 * @property category
	 * @since 0.7.0
	 */
	public var category: Category
		private set

	/**
	 * The source's location.
	 * @property location
	 * @since 0.7.0
	 */
	public var location: String
		private set

	/**
	 * The source's data.
	 * @property data
	 * @since 0.7.0
	 */
	public val data: String by lazy {
		this.read()
	}

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 * @hidden
	 */
	init {
		this.context = context
		this.location = location
		this.category = category
	}

	/**
	 * @method read
	 * @since 0.7.0
	 * @hidden
	 */
	private fun read(): String {

		try {

			if (this.location.startsWith("http://") ||
				this.location.startsWith("https://")) {
				return RemoteFileLoader().execute(this.location).get()
			}

			return this.context.assets.open(this.location).reader().use {
				it.readText()
			}

		} catch (e: IOException) {
			throw Exception("Cannot load source at location ${this.location}.")
		}
	}
}

File: platforms/ios/Dezel/Dezel/Application/Source.swift
import Foundation

/**
 * @class Source
 * @since 0.7.0
 */
open class Source {

	//--------------------------------------------------------------------------
	// MARK: Enum
	//--------------------------------------------------------------------------

	/**
	 * @enum Type
	 * @since 0.7.0
	 */
	public enum `Type` {
		case style
		case script
	}

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * @property type
	 * @since 0.7.0
	 */
	private(set) public var type: Type

	/**
	 * @property path
	 * @since 0.7.0
	 */
	private(set) public var path: String

	/**
	 * @property data
	 * @since 0.7.0
	 */
	public lazy var data: String = {
		return self.read()
	}()

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
	 * @constructor
	 * @since 0.7.0
	 */
	public init(type: Type, path: String) {
		self.type = type
		self.path = path
	}

	//--------------------------------------------------------------------------
	// MARK: Private API
	//--------------------------------------------------------------------------

	/**
	 * @method read
	 * @since 0.7.0
	 * @hidden
	 */
	private func read() -> String{

		do {

			if (self.path.hasPrefix("http://") ||
				self.path.hasPrefix("https://")) {
				return try String(contentsOf: URL(string: self.path)!)
			}

			return try String(contentsOfFile: Bundle.main.path(forResource: self.path, ofType: nil)!)

		} catch _ {
			fatalError("Cannot load source at location \(self.path).")
		}
	}
}
